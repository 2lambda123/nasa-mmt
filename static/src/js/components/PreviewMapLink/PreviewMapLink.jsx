import React from 'react'
import PropTypes from 'prop-types'

class PreviewMapLink extends React.Component {
  getMapQuery(data, type) {
    let query = ''
    if (type === 'point') {
      if (data.Longitude && data.Latitude) {
        query = `${data.Longitude},${data.Latitude}`
      }

      if (query.length) {
        query = `sp=${encodeURIComponent(query)}`
      }

      return query
    }

    if (type === 'bounding_rectangle') {
      if (
        data.WestBoundingCoordinate
        && data.SouthBoundingCoordinate
        && data.EastBoundingCoordinate
        && data.NorthBoundingCoordinate
      ) {
        query = `${data.WestBoundingCoordinate},${data.SouthBoundingCoordinate},${data.EastBoundingCoordinate},${data.NorthBoundingCoordinate}`
      }

      if (query.length) {
        query = `sb=${encodeURIComponent(query)}`
      }

      return query
    }

    if (type === 'polygon') {
      const points = data.Boundary.Points
      points.forEach((point) => {
        if (point.Longitude && point.Latitude) {
          if (query.length) {
            query = `${query},`
          }

          query = `${query}${point.Longitude},${point.Latitude}`
        }
      })

      if (query.length) {
        query = `polygon=${encodeURIComponent(query)}`
      }

      return query
    }

    return ''
  }

  render() {
    const { type, formData } = this.props
    let url = process.env.REACT_APP_PREVIEW_MAP_URL ? process.env.REACT_APP_PREVIEW_MAP_URL : 'https://search.earthdata.nasa.gov/search/map'
    const mapQuery = this.getMapQuery(formData, type)
    if (mapQuery.length) {
      url = `${url}?${mapQuery}`
    }

    return (
      <div>
        <a href={url} target="_blank" rel="noopener noreferrer">
          <i className="eui-icon eui-fa-globe" />
          {' '}
          Preview on Map
        </a>
      </div>
    )
  }
}

PreviewMapLink.propTypes = {
  type: PropTypes.string.isRequired,
  formData: PropTypes.shape({
    Longitude: PropTypes.number,
    Latitude: PropTypes.number,
    WestBoundingCoordinate: PropTypes.number,
    SouthBoundingCoordinate: PropTypes.number,
    EastBoundingCoordinate: PropTypes.number,
    NorthBoundingCoordinate: PropTypes.number,
    Boundary: PropTypes.shape({
      Points: PropTypes.arrayOf(
        PropTypes.shape({
          Longitude: PropTypes.number.isRequired,
          Latitude: PropTypes.number.isRequired
        })
      )
    })
  }).isRequired
}

export default PreviewMapLink
