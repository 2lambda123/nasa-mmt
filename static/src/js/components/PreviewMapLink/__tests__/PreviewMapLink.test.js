import React from 'react'
import { render } from '@testing-library/react'
import PreviewMapLink from '../PreviewMapLink'

describe('PreviewMapLink Test', () => {
  it('link to preview map with a point', async () => {
    const props = {
      formData: {
        Longitude: 10,
        Latitude: 20
      }
    }
    const { container } = render(
      <PreviewMapLink {...props} type="point" />
    )
    expect(document.querySelector('a').getAttribute('href')).toBe('https://search.earthdata.nasa.gov/search/map?sp=10%2C20')
    expect(container).toMatchSnapshot()
  })

  it('link to preview map with a bounding rectangle', async () => {
    const props = {
      formData: {
        WestBoundingCoordinate: -40,
        SouthBoundingCoordinate: 10,
        EastBoundingCoordinate: 50,
        NorthBoundingCoordinate: 70
      }
    }
    const { container } = render(
      <PreviewMapLink {...props} type="bounding_rectangle" />
    )
    expect(document.querySelector('a').getAttribute('href')).toBe('https://search.earthdata.nasa.gov/search/map?sb=-40%2C10%2C50%2C70')
    expect(container).toMatchSnapshot()
  })

  it('link to preview map with a polygon', async () => {
    const props = {
      formData: {
        Boundary: {
          Points: [
            {
              Longitude: -30,
              Latitude: 50
            },
            {
              Longitude: 10,
              Latitude: 40
            },
            {
              Longitude: 35,
              Latitude: 50
            },
            {
              Longitude: -30,
              Latitude: 50
            },
          ]
        }
      }
    }
    const { container } = render(
      <PreviewMapLink {...props} type="polygon" />
    )
    expect(document.querySelector('a').getAttribute('href')).toBe('https://search.earthdata.nasa.gov/search/map?polygon=-30%2C50%2C10%2C40%2C35%2C50%2C-30%2C50')
    expect(container).toMatchSnapshot()
  })

  it('link to preview map without parameters', async () => {
    const props = {
      formData: {
      }
    }
    const { container } = render(
      <PreviewMapLink {...props} type="bounding_rechtangle" />
    )
    expect(document.querySelector('a').getAttribute('href')).toBe('https://search.earthdata.nasa.gov/search/map')
    expect(container).toMatchSnapshot()
  })
})
