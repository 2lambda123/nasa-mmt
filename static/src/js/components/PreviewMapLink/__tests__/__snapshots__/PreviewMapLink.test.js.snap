// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewMapLink Test link to preview map with a bounding rectangle 1`] = `
<div>
  <div>
    <a
      href="https://search.earthdata.nasa.gov/search/map?sb=-40%2C10%2C50%2C70"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        class="eui-icon eui-fa-globe"
      />
       
      Preview on Map
    </a>
  </div>
</div>
`;

exports[`PreviewMapLink Test link to preview map with a point 1`] = `
<div>
  <div>
    <a
      href="https://search.earthdata.nasa.gov/search/map?sp=10%2C20"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        class="eui-icon eui-fa-globe"
      />
       
      Preview on Map
    </a>
  </div>
</div>
`;

exports[`PreviewMapLink Test link to preview map with a polygon 1`] = `
<div>
  <div>
    <a
      href="https://search.earthdata.nasa.gov/search/map?polygon=-30%2C50%2C10%2C40%2C35%2C50%2C-30%2C50"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        class="eui-icon eui-fa-globe"
      />
       
      Preview on Map
    </a>
  </div>
</div>
`;
